infoschema/tables_test.go

```diff
@@ -1366,3 +1367,94 @@ func (s *testTableSuite) TestServerInfoResolveLoopBackAddr(c *C) {
 		c.Assert(n.StatusAddr, Equals, "192.168.130.22:10080")
 	}
 }
+
+func (s *testTableSuite) TestPlacementPolicy(c *C) {
+	tk := s.newTestKitWithRoot(c)
+	tk.MustExec("use test")
+	tk.MustExec("create table test_placement(id int primary key) partition by hash(id) partitions 2")
+
+	is := s.dom.InfoSchema()
+	tb, err := is.TableByName(model.NewCIStr("test"), model.NewCIStr("test_placement"))
+	c.Assert(err, IsNil)
+	partDefs := tb.Meta().GetPartitionInfo().Definitions
+
+	bundles := make(map[string]*placement.Bundle)
+	is.MockBundles(bundles)
+	tk.MustQuery("select group_id, group_index, rule_id, schema_name, table_name, partition_name, index_name, " +
+		"role, replicas, constraints from information_schema.placement_policy").Check(testkit.Rows())
+
+	bundleID := "pd"
+	bundle := &placement.Bundle{
+		ID:       bundleID,
+		Index:    0,
+		Override: false,
+		Rules: []*placement.Rule{
+			{
+				GroupID: bundleID,
+				ID:      "default",
+				Role:    "voter",
+				Count:   3,
+			},
+		},
+	}
+	bundles[bundleID] = bundle
+	tk.MustQuery("select * from information_schema.placement_policy").Check(testkit.Rows())
+
+	bundleID = fmt.Sprintf("%s%d", placement.BundleIDPrefix, partDefs[0].ID)
+	bundle = &placement.Bundle{
+		ID:       bundleID,
+		Index:    3,
+		Override: true,
+		Rules: []*placement.Rule{
+			{
+				GroupID: bundleID,
+				ID:      "0",
+				Role:    "voter",
+				Count:   3,
+				LabelConstraints: []placement.LabelConstraint{
+					{
+						Key:    "zone",
+						Op:     "in",
+						Values: []string{"bj"},
+					},
+				},
+			},
+		},
+	}
+	bundles[bundleID] = bundle
+	expected := fmt.Sprintf("%s 3 0 test test_placement p0 <nil> voter 3 \"+zone=bj\"", bundleID)
+	tk.MustQuery("select group_id, group_index, rule_id, schema_name, table_name, partition_name, index_name, " +
+		"role, replicas, constraints from information_schema.placement_policy").Check(testkit.Rows(expected))
+
+	bundle.Rules[0].LabelConstraints[0].Values = []string{"bj", "sh"}
+	tk.MustQuery("select constraints from information_schema.placement_policy").Check(testkit.Rows("\"+zone=bj,+zone=sh\""))
+
+	bundle.Rules[0].LabelConstraints = []placement.LabelConstraint{
+		{
+			Key:    "zone",
+			Op:     "in",
+			Values: []string{"bj"},
+		},
+		{
+			Key:    "zone",
+			Op:     "notIn",
+			Values: []string{"sh"},
+		},
+	}
+	tk.MustQuery("select constraints from information_schema.placement_policy").Check(testkit.Rows("\"+zone=bj\",\"-zone=sh\""))
+
+	rule1 := bundle.Rules[0].Clone()
+	rule1.ID = "1"
+	bundle.Rules = append(bundle.Rules, rule1)
+	tk.MustQuery("select rule_id, schema_name, table_name, partition_name from information_schema.placement_policy order by rule_id").Check(testkit.Rows(
```

Comment:
if you drop a partitioning rule but it hasn't been cleaned by GC, could it be get here?

Reply:
I didn't really add or drop a partition, so GC won't clean the rule.

Reply:
It is #20622 to handle this case smoothly. It will drop the cache, but not to inform PD. So we can recover table/partitions from PD, or wait GC to delete rules on PD.

Reply:
this may be better for the show result